src/ray/gcs/pubsub/test/gcs_pub_sub_test.cc

```diff
@@ -0,0 +1,322 @@
+    client_ = std::make_shared<gcs::RedisClient>(redis_client_options);
+    RAY_CHECK_OK(client_->Connect(io_service_));
+  }
+
+  virtual void TearDown() override {
+    io_service_.stop();
+    thread_io_service_->join();
+  }
+
+  template <typename PUB_SUB, typename ID, typename Data>
+  void Subscribe(PUB_SUB &pub_sub, const std::string &channel, const ID &id,
+                 std::vector<Data> &result) {
+    std::promise<bool> promise;
+    auto done = [&promise](Status status) { promise.set_value(status.ok()); };
+    auto subscribe = [&result](const ID &id, const Data &data) {
+      result.push_back(data);
+    };
+    RAY_CHECK_OK(pub_sub.Subscribe(channel, id, subscribe, done));
+    WaitReady(promise.get_future(), timeout_ms_);
+  }
+
+  template <typename PUB_SUB, typename ID, typename Data>
+  void SubscribeAll(PUB_SUB &pub_sub, const std::string &channel,
+                    std::vector<std::pair<ID, Data>> &result) {
+    std::promise<bool> promise;
+    auto done = [&promise](Status status) { promise.set_value(status.ok()); };
+    auto subscribe = [&result](const ID &id, const Data &data) {
+      result.push_back(std::make_pair(id, data));
+    };
+    RAY_CHECK_OK(pub_sub.SubscribeAll(channel, subscribe, done));
+    WaitReady(promise.get_future(), timeout_ms_);
+  }
+
+  template <typename PUB_SUB, typename ID>
+  bool Unsubscribe(PUB_SUB &pub_sub, const std::string &channel, const ID &id) {
+    std::promise<bool> promise;
+    auto done = [&promise](Status status) { promise.set_value(status.ok()); };
+    RAY_CHECK_OK(pub_sub.Unsubscribe(channel, id, done));
+    return WaitReady(promise.get_future(), timeout_ms_);
+  }
+
+  template <typename PUB_SUB, typename ID, typename Data>
+  bool Publish(PUB_SUB &pub_sub, const std::string &channel, const ID &id,
+               const Data &data) {
+    std::promise<bool> promise;
+    auto done = [&promise](Status status) { promise.set_value(status.ok()); };
+    RAY_CHECK_OK(pub_sub.Publish(channel, id, data, done));
+    return WaitReady(promise.get_future(), timeout_ms_);
+  }
+
+  bool WaitReady(std::future<bool> future, const std::chrono::milliseconds &timeout_ms) {
+    auto status = future.wait_for(timeout_ms);
+    return status == std::future_status::ready && future.get();
+  }
+
+  template <typename Data>
+  void WaitPendingDone(const std::vector<Data> &data, int expected_count) {
+    auto condition = [&data, expected_count]() {
+      return (int)data.size() == expected_count;
+    };
+    EXPECT_TRUE(WaitForCondition(condition, timeout_ms_.count()));
+  }
+
+  std::shared_ptr<gcs::RedisClient> client_;
+  JobID job_id_ = JobID::FromInt(1);
+  TaskID task_id_ = TaskID::ForDriverTask(job_id_);
+  ClientID node_id_ = ClientID::FromRandom();
+  const std::chrono::milliseconds timeout_ms_{10000};
+
+ private:
+  boost::asio::io_service io_service_;
+  std::unique_ptr<std::thread> thread_io_service_;
+};
+
+TEST_F(GcsPubSubTest, TestJobTablePubSubApi) {
```

Comment:
I think it's okay to not test every table in the unit tests. 
The implementation of `GcsPubSub` should be independent of what data is being pubsub-ed.
More importantly, we should test the functionalities of `GcsPubSub` class in various scenarios, like multi-threading, subscribe then unsubscribe then subscribe again, etc.

Reply:
fixed